core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -159,63 +176,86 @@ public String getRegistry() {
+        return asCompatibleSubstituteFor(DockerImageName.parse(otherImageName));
     }
 
-    @Data
-    private static class TagVersioning implements Versioning {
-        public static final String TAG_REGEX = "[\\w][\\w\\.\\-]{0,127}";
-        private final String tag;
-
-        TagVersioning(String tag) {
-            this.tag = tag;
-        }
-
-        @Override
-        public boolean isValid() {
-            return tag.matches(TAG_REGEX);
-        }
-
-        @Override
-        public String getSeparator() {
-            return ":";
-        }
-
-        @Override
-        public String toString() {
-            return tag;
-        }
+    /**
+     * Declare that this {@link DockerImageName} is a compatible substitute for another image - i.e. that this image
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(DockerImageName otherImageName) {
+        return new DockerImageName(rawName, registry, repo, versioning, otherImageName);
     }
 
-    @Data
-    private static class Sha256Versioning implements Versioning {
-        public static final String HASH_REGEX = "[0-9a-fA-F]{32,}";
-        private final String hash;
-
-        Sha256Versioning(String hash) {
-            this.hash = hash;
+    /**
+     * Test whether this {@link DockerImageName} has declared compatibility with another image (set using
+     * {@link DockerImageName#asCompatibleSubstituteFor(String)} or
+     * {@link DockerImageName#asCompatibleSubstituteFor(DockerImageName)}.
+     * <p>
+     * If a version tag part is present in the <code>other</code> image name, the tags must exactly match, unless it
+     * is 'latest'. If a version part is not present in the <code>other</code> image name, the tag contents are ignored.
+     *
+     * @param other the other image that we are trying to test compatibility with
+     * @return whether this image has declared compatibility.
+     */
+    public boolean isCompatibleWith(DockerImageName other) {
+        // is this image already the same?
+        final boolean thisRegistrySame = other.registry.equals(this.registry);
+        final boolean thisRepoSame = other.repo.equals(this.repo);
+        final boolean thisVersioningNotSpecifiedOrSame = other.versioning == null ||
+            other.versioning.equals(Versioning.TagVersioning.LATEST) ||
+            other.versioning.equals(this.versioning);
+
+        if (thisRegistrySame && thisRepoSame && thisVersioningNotSpecifiedOrSame) {
+            return true;
         }
 
-        @Override
-        public boolean isValid() {
-            return hash.matches(HASH_REGEX);
+        if (this.compatibleSubstituteFor == null) {
```

Comment:
idea: make `compatibleSubstituteFor` `@NonNull`, use `SelfCompatible` by default

Reply:
I didn't really manage to make this work, or at least elegantly.